Consul: Update integrate/vault/k8s for v1.21 and v1.22 (#1560)
hashicorp/web-unified-docs: content/consul/v1.21.x/content/docs/integrate/vault/k8s.mdx

# Use Vault for secrets management with Consul on Kubernetes

This page describes how to securely deploy Consul on Kubernetes using Vault as a secrets manager. Using Vault, you can securely store and centrally manage the secrets you will need to provide to your Consul agents. These include, but are not limited to, gossip encryption keys, TLS certificates, service mesh certificates. Furthermore, you can also [store and manage your Consul Enterprise license](/consul/docs/deploy/server/k8s/vault/data/enterprise-license).

The following architecture diagram depicts the desired outcome.

![Architectural diagram for Vault as Consul secrets manager](/img/kubernetes-diagram-vault-as-secrets-manager.png)

## Workflow

The following steps describe specific tasks for Vault secrets management for Consul on Kubernetes:

> 1. Store enterprise license in Vault

Reviewer: ```suggestion
- Store Enterprise license in Vault
```
This list should be an unordered list, and each line item should link to the section in the docs below.

```yaml
global:
##...
  enterpriseLicense:
    secretName: 'consul/data/secret/enterpriselicense'
    secretKey: 'key'
##...
```

### Store Consul gossip key in Vault

First, generate and store the encryption key in Vault.

```shell-session
$ vault kv put consul/secret/gossip gossip="$(consul keygen)"
Key              Value
---              -----
created_time     2022-03-16T18:18:52.389731147Z
deletion_time    n/a
destroyed        false
version          1
```

Define a policy that grants access to the path where you stored the gossip encryption.

```shell-session
$ vault policy write gossip-policy - <<EOF
path "consul/data/secret/gossip" {
  capabilities = ["read"]
}
EOF
```

The command should return an output similar to the following.

<CodeBlockConfig hideClipboard>

```plaintext
Success! Uploaded policy: gossip-policy
```

</CodeBlockConfig>

Then, edit the Consul Helm chart values file to reference the Vault path and key where the gossip key is stored.

```yaml
global:
  gossipEncryption:
    secretName: consul/data/secret/gossip
    secretKey: gossip
```

> ### Setup PKI secrets engine for TLS and service mesh CA

Reviewer: ```suggestion
## Setup PKI secrets engine for TLS and service mesh CA
```

# Use Vault for secrets management with Consul on Kubernetes

This page describes how to securely deploy Consul on Kubernetes using Vault as a secrets manager. Vault can securely store and centrally manage the secrets you will need to provide to your Consul agents. These  secrets include, but are not limited to, gossip encryption keys, TLS certificates, service mesh certificates. 

> You can also use Vault to [store and manage your Consul Enterprise license](/consul/docs/deploy/server/k8s/vault/data/enterprise-license).

Reviewer: ```suggestion
For more information about deploying Vault with Consul on Kubernetes, refer to [Vault as secrets backend](/consul/docs/deploy/server/k8s/vault).
```
My suggestion to account for the other existing content. This page is an introduction to using Vault for Consul on Kubernetes, then link to the section with more comprehensive information.

</CodeBlockConfig>

Next, enable Vault's PKI secrets engine at the `connect-root` path to be used as root CA for Consul service mesh.

```shell-session
$ vault secrets enable -path connect-root pki
Success! Enabled the pki secrets engine at: connect-root/
```

> #### Consul server policy

Reviewer: ```suggestion
### Consul server policy
```

> #### Consul CA access policy

Reviewer: ```suggestion
### Consul CA access policy
```

</CodeBlockConfig>

#### Configure Kubernetes authentication

> Vault provides a [Kubernetes authentication](/vault/docs/auth/kubernetes) method that enables clients to authenticate with a Kubernetes Service Account Token. Using the Kubernetes authentication, Vault identifies your Consul agents using their service account and gives them access to the secrets they need to join your Consul datacenter.

Reviewer: ```suggestion
Vault provides a [Kubernetes authentication method](/vault/docs/auth/kubernetes) that enables clients to authenticate with a Kubernetes Service Account Token. Using the Kubernetes authentication, Vault identifies your Consul agents using their service account and gives them access to the secrets they need to join your Consul datacenter.
```

</CodeBlockConfig>

The following diagram provides a summary of the configuration you created.

![Permissions for Vault as Consul secrets manager](/img/kubernetes-diagram-permissions-vault-as-secrets-manager.png)

> #### Configure Consul Helm chart to use Vault as secrets manager

Reviewer: ```suggestion
### Configure Consul Helm chart to use Vault as secrets manager
```
